File: plugins/java-code-generator/src/main/java/io/samjingwen/codegenerator/CodeGeneratorPlugin.java
package io.samjingwen.codegenerator;

import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.tasks.SourceSet;
import org.gradle.api.tasks.SourceSetContainer;

public abstract class CodeGeneratorPlugin implements Plugin<Project> {

  @Override
  public void apply(Project project) {
    if (!project.getPluginManager().hasPlugin("java")) {
      throw new IllegalStateException("Java plugin is required!");
    }

    var sourceSets = project.getExtensions().getByType(SourceSetContainer.class);
    var mainSourceSet = sourceSets.getByName(SourceSet.MAIN_SOURCE_SET_NAME);

    var buildDir = project.getBuildDir().getPath();
    mainSourceSet.getJava().srcDir(buildDir + "/generated/codegen/src/main/java");

    var tasks = project.getTasks();
    var codeGenerationTasks = tasks.withType(CodeGenerationTask.class);

    tasks
        .named("compileJava")
        .configure(compileJavaTask -> compileJavaTask.dependsOn(codeGenerationTasks));
  }
}

File: plugins/code-generator/src/main/kotlin/io/samjingwen/codegenerator/CodeGeneratorPlugin.kt
package io.samjingwen.codegenerator

import org.gradle.api.Plugin
import org.gradle.api.Project

class CodeGeneratorPlugin : Plugin<Project> {
    override fun apply(project: Project) {
        val extension = project.extensions.create("generateCode", CodeGeneratorExtension::class.java)
        project.tasks.register("generateCode", CodeGenerationTask::class.java) {
            it.templateFile.set(extension.templateFile)
            it.sourceFile.set(extension.sourceFile)
            it.outputFile.set(extension.outputFile)
        }
    }
}
